tap-version-14-specification.md

```diff
@@ -0,0 +1,741 @@
+`\n`.
+
+TAP14's general format is:
+
+```
+TAPDocument ->
+  Version Plan Body |
+  Version Body Plan
+```
+
+The Version is always `TAP version 14`.
+
+The Body is a collection of lines representing a test set.
+
+The Plan reports on the number of tests included in the Body.
+
+Any line matching the pattern `/^\s*#.*$/` is treated as a diagnostic
+comment, and MAY be reported to the user or ignored, but MUST be
+ignored for the purposes of interpreting test output semantics.
+
+For example:
+
+```
+TAP version 14
+1..4
+ok 1 Description # Directive
+# Diagnostic
+  ---
+  message: 'Failure message'
+  severity: fail
+  data:
+    found:
+      - 1
+      - 3
+      - 2
+    wanted:
+      - 1
+      - 2
+      - 3
+  ...
+ok 2 Description
+ok 3 Description
+    # Subtest: name of the child test
+    ok 1 this is ok
+    ok 2 this is also ok
+    not ok 3 this is a failure
+      ---
+      message: 'Child Failure'
+      severity: fail
+      ...
+    1..3
+not ok 4 name of the child test
+```
+
+For example, a test file's output might look like:
+
+```
+TAP version 14
+1..4
+ok 1 - Input file opened
+not ok 2 - First line of the input valid
+  ---
+  message: 'First line invalid'
+  severity: fail
+  data:
+    got: 'Flirble'
+    expect: 'Fnible'
+  ...
+ok 3 - Read the rest of the file
+not ok 4 - Summarized correctly # TODO Not written yet
+  ---
+  message: "Can't make summary yet"
+  severity: todo
+  ...
+```
+
+## HARNESS BEHAVIOR
+
+In this document, the "harness" is any program analyzing TAP output.
+
+A harness interpreting a subprocess MUST only read TAP output from
+standard output and not from standard error.
+
+Lines written to standard output matching `/^(not)? ok\b/` must be
+interpreted as test lines. After a test line a block of lines starting
+with `/^ {2}---$/` and ending with `/^ {2}\.\.\.$/` will be interpreted as an
+inline YAML document providing extended diagnostic information about
+the preceding test.
+
+All other lines must not be considered test output.
+
+## TESTS LINES AND THE PLAN
+
+### The Version
+
+To indicate that this is TAP14 the first non-comment, non-whitespace
+line must be:
+
+```
+TAP version 14
+```
+
+A compliant TAP14 Harness SHOULD interpret TAP streams with `TAP
+version 13` according to TAP13 specification.
+
+TAP streams lacking a version number SHOULD be interpreted according to
+the TAP12 specification.
+
+In order to facilitate forward compatibility, a compliant TAP14
+Harness MAY interpret test data according to TAP14, regardless of
+stated version number.  For example, it may accept subtests in TAP12
+or TAP13 streams, even though these were not supported in those
+versions of TAP.
+
+Any stated version number less than 13 is an error.
+
+### The Plan
+
+The Plan tells how many tests will be run, or how many tests have run.
+It's a check that the test file hasn't stopped prematurely.
+
+The Plan MUST appear exactly once, EITHER:
```

Comment:
Is there a valid use case for multiple plan lines? A few weeks ago, I was testing the streaming interface for the Python TAP consumer that I've written and was surprised when `prove -v` provided multiple plan lines. It caused my consumer to report some errors because of this constraint to 1 appearance of the plan.

Initially I thought that `prove` might be deficient as a TAP producer, but the behavior seemed very reasonable the more that I thought about it. I've encountered a number of TAP producer libraries that specify a plan per file.

Maybe the question really should be: should TAP expect producers to aggregate plan counts into a single plan or should TAP consumers be able to deal with multiple plans from independent streams?

One option for supporting multiple plans would be a way to demarcate the end of a stream. Perhaps something matching `/^EOTAP$/`.